Core/include/Acts/Navigation/NavigationStateUpdaters.hpp

```diff
@@ -39,31 +39,25 @@ inline void updateCandidates(const GeometryContext& gctx,
   const auto& position = nState.position;
   const auto& direction = nState.direction;
 
-  NavigationState::SurfaceCandidates nextSurfaceCandidates;
-
-  for (NavigationState::SurfaceCandidate c : nState.surfaceCandidates) {
+  for (NavigationState::SurfaceCandidate& c : nState.surfaceCandidates) {
     // Get the surface representation: either native surface of portal
-    const Surface& sRep =
+    const Surface& surface =
         c.surface != nullptr ? *c.surface : c.portal->surface();
 
     // Only allow overstepping if it's not a portal
     ActsScalar overstepTolerance =
         c.portal != nullptr ? s_onSurfaceTolerance : nState.overstepTolerance;
 
-    // Get the intersection @todo make a templated intersector
-    // TODO surface tolerance
-    auto sIntersection = sRep.intersect(
+    // Check the surface intersection
+    auto sIntersection = surface.intersect(
         gctx, position, direction, c.boundaryTolerance, s_onSurfaceTolerance);
     for (auto& si : sIntersection.split()) {
```

Comment:
the splitting should not be necessary if we knew already with intersection index we aim for in case of an `intersection candidate`

Reply:
correct, the only issue would be if a formally non-reachable split intersection becomes reachable right?

I thought about it a lot (too much) yesterday, but probably the solution is the following:

- get the list of surfaces (and keep them separate from the intersection)
- have a buffer of intersection solutions on the candidates and remember the actual buffer size (which can change with re-navigation)

Reply:
Yes I think this is similar to what I do in the try all navigators. I don't really see a way around this having surfaces which can be intersected twice.

Reply:
In the Gen1 navigator I disambiguated that by having a vector of `SurfaceIntersections` as candidates. These can also be used to reintersect.